miner/active_set_generator.go

```diff
@@ -0,0 +1,320 @@
+	}
+
+	// we use it to avoid running `prepare` method in parallel
+	linearized sync.Mutex
+}
+
+func (p *activeSetGenerator) updateFallback(target types.EpochID, set []types.ATXID) {
+	p.log.Info("received trusted activeset update",
+		target.Field().Zap(),
+		zap.Int("size", len(set)),
+	)
+	p.fallback.Lock()
+	defer p.fallback.Unlock()
+	if _, exists := p.fallback.data[target]; exists {
+		p.log.Debug("fallback active set already exists", target.Field().Zap())
+		return
+	}
+	p.fallback.data[target] = set
+}
+
+// ensure tries to generate active set for the target epoch in configured number of tries,
+// each try will be repeated after configured retry interval.
+func (a *activeSetGenerator) ensure(ctx context.Context, target types.EpochID) {
+	var err error
+	for try := 0; try < a.cfg.activeSet.Tries; try++ {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(a.cfg.activeSet.RetryInterval):
+		}
+		current := a.clock.CurrentLayer()
+		// we run it here for side effects
+		_, _, _, err = a.generate(current, target)
+		if err == nil {
+			return
+		}
+		a.log.Debug("failed to prepare active set", zap.Error(err), zap.Uint32("attempt", uint32(try)))
+	}
+	a.log.Warn("failed to prepare active set", zap.Error(err))
+}
+
+// generate generates activeset.
+//
+// It persists it in the local database, so that when node is restarted
+// it doesn't have to redo the work.
+//
+// The method is expected to be called at any point in current epoch, and at the very end of the previous epoch.
+func (p *activeSetGenerator) generate(
+	current types.LayerID,
+	target types.EpochID,
+) (types.Hash32, uint64, []types.ATXID, error) {
+	p.linearized.Lock()
+	defer p.linearized.Unlock()
+
+	id, setWeight, set, err := activeset.Get(p.localdb, activeset.Tortoise, target)
+	if err != nil && !errors.Is(err, sql.ErrNotFound) {
+		return id, 0, nil, fmt.Errorf("failed to get prepared active set: %w", err)
+	}
+	if err == nil {
+		return id, setWeight, set, nil
+	}
+
+	p.fallback.Lock()
+	fallback, exists := p.fallback.data[target]
+	p.fallback.Unlock()
+
+	start := time.Now()
+	if exists {
+		p.log.Info("generating activeset from trusted fallback",
+			target.Field().Zap(),
+			zap.Int("size", len(fallback)),
+		)
+		var err error
+		setWeight, err = getSetWeight(p.atxsdata, target, fallback)
+		if err != nil {
+			return id, 0, nil, err
+		}
+		set = fallback
+	} else {
+		epochStart := p.clock.LayerToTime(target.FirstLayer())
+		networkDelay := p.cfg.networkDelay
+		p.log.Info("generating activeset from grades", target.Field().Zap(),
+			zap.Time("epoch start", epochStart),
+			zap.Duration("network delay", networkDelay),
+		)
+		result, err := activeSetFromGrades(p.db, target, epochStart, networkDelay)
+		if err != nil {
+			return id, 0, nil, err
+		}
+		if result.Total == 0 {
+			return id, 0, nil, fmt.Errorf("empty active set")
+		}
+		if result.Total > 0 && len(result.Set)*100/result.Total > p.cfg.goodAtxPercent {
```

Comment:
The `result.Total > 0` is already satisfied by the check above.
```suggestion
		if result.Total == 0 {
			return id, 0, nil, fmt.Errorf("empty active set")
		}
		if len(result.Set)*100/result.Total > p.cfg.goodAtxPercent {
```